dame_flame/grouped_mr.py

```diff
@@ -51,18 +51,21 @@ def algo2_GroupedMR(df_all, df_unmatched, covs_match_on, all_covs, treatment_col
 
     # These are the unique values in the bi col. length = number of groups
     unique_matched_row_vals = np.unique(bi)
-        
+    
+    # This list stores all matched groups created by this function
+    all_units_in_g = []
     for bi_val in unique_matched_row_vals:
         # type "int64index", ~ list, all of the unit_numbers in a matched group.
         units_in_g = matched_rows.index[matched_rows['b_i']==bi_val]
-        
+        # Append newly created matched group
+        all_units_in_g.append(list(units_in_g))
```

Comment:
Found the issue you were encountering on sample.csv (I'm pretty sure, but do double check me on this). 

So that dataset first dropped x4 and found a group of units 0,1. Cool. 
Then on the second iteration, it's here and considering dropping x3. When we look at who we can match on x3&x4, then units 0&1 match again and form a group again. So, 0,1 is units_in_g. But, the line below that is where it checks to see if any of the units in the group [0,1] already have groups formed, and the if statement below that is where it makes sure it only adds to return_groups (the listing with the "*"s in the event that a new MMG is being created. 

So, you'll want to move the line where you add to all_units_in_g after the `if len(newly_matched) !=0`. The presence of that if-statement is why return_groups only had one entry for 0&1 but your units_in_g had multiple entries of those. 

(I suggest to debug this by sticking a print statement in for units_in_g and newly_matched right when they're first created)

Reply:
That is totally the issue. Thanks for catching this!